### PR TITLE
ng_pktbuf: fix possible packet corruption in case of internal error

### LINKED ISSUE #3365: ng_pktbuf_add: possible packet corruption in case of internal error
[Link to source](https://github.com/RIOT-OS/RIOT/blob/d2d4e69d8acdd4988d06917b7d89e1aa0e1b97cd/sys/net/crosslayer/ng_pktbuf/ng_pktbuf.c#L275)

```c
pkt->size -= size;
DEBUG("pktbuf: resize pkt->size to %u\n", (unsigned)pkt->size);
pkt->data = (void *)(((uint8_t *)pkt->data) + size);
```
...
```c
if (!_pktbuf_internal_add_pkt(pkt->data)) {
    DEBUG("failed (freeing %p)\n", (void *)new_pktsnip);
    _pktbuf_internal_free(new_pktsnip);

    return NULL;
}
```

In case this fails it leave `pkt` altered, i.e. `pkt->size -= size` and `pkt->data` changed.

The only thing is that I don't really understand the circumstances for ` _pktbuf_internal_add_pkt(pkt->data) == false`. I get this when I supply `size` so that `size == pkt->size`, i.e. trying to dissect a pktbuf in one part that has all the data and one part that has `size = 0`.

### PR DESCRIPTION
**Pull request by authmillenon**
Fixes #3365

### CODE REVIEW

File: sys/net/crosslayer/ng_pktbuf/ng_pktbuf.c [hunk 1]
```diff
@@ -258,6 +258,16 @@ static ng_pktsnip_t *_pktbuf_add_unsafe(ng_pktsnip_t *pkt, void *data,
             return NULL;
         }
 
+        DEBUG("pktbuf: Adding chunk to %p ", pkt->data);
+
+        if (!_pktbuf_internal_add_pkt(pkt->data)) {
```
Comment (daniel-k) [inline sys/net/crosslayer/ng_pktbuf/ng_pktbuf.c]: Could you please explain what this check is meant to do? Yesterday I stared at it for about 30min but I'm still not sure why you need to perform that.

On the other hand this now checks `pkt->data` before it is incremented. On purpose?

Comment (authmillenon) [inline sys/net/crosslayer/ng_pktbuf/ng_pktbuf.c]: > On the other hand this now checks pkt->data before it is incremented. On purpose?

Makes no different. The pointer only needs to be an address in the chunk that is representing the packet internally.

**Comment by daniel-k** [comment]
Works as predicted in #3365 and also fixes my problem. Thank you very much :)

ACK

**Comment by daniel-k** [comment]
@authmillenon 

Unfortunately this introduces a new bug that leads to a memory leak. It happens when `pkt_buf_add()` resizes pktsnips to `size = 0` (which is exactly my use case from above).

Short sketch of what happens:

1. You have pktsnip X of type `type_A` that holds 4 byte of data @ 0x100
2. `ng_pktbuf_add(X, X->data, 4, type_B)` "dissecting"
3. You get: `X->type = type_A`, `X->size = 0` | 
    `X->next->type = type_B`, `X->next->size = 4`
4. So far so good, **but** the node (`_used_t*`) that holds `X->next->data` now has `node->pkts = 2`. `X->data` (the one with size 0) won't be freed by `_pktbuf_internal_free` because it points outside the buffer, so `node->pkts` will never be 0 and therefore not freed. Voilà, the memory leak.

This all boils down to bad handling of `size = 0` pktsnips or rather wrong handling in `_pktbuf_add_unsafe(...)`.

I didn't create a PR because I don't know how you want to handle pktsnips with `size = 0`, so how to proceed? Revert for now?

**Comment by authmillenon** [comment]
Ah… I meant to fix that but somehow forgot about it. Does #3374 do it for you?

**Pull request merged**